File: ios/Sources/Spike/SpikeScene.swift
import Observation
import RealityKit
import SwiftUI
import UIKit

/// What the SwiftUI layer reads from the scene: the button's projected rectangle, for the
/// accessibility overlay.
@MainActor @Observable
final class SpikeOverlay {
    var buttonRect: CGRect?
}

/// The renderer spike's vertical slice (SMASH-2), the twin of Android's SpikeScene: the Oasis
/// world, a floating title, a camera-parented score HUD and one 3D button that squashes on a
/// spring. Every placement number mirrors Android's so the two can be laid side by side.
@MainActor
final class SpikeScene {
    let root = Entity()
    let overlay = SpikeOverlay()
    let buttonLabelText = String(localized: "play.button")
    var viewSize = CGSize(width: 1, height: 1)
    var reduceMotion = false

    private let camera = PerspectiveCamera()
    private let fovDegrees: Float = 50
    private let hudDepth: Float = 4
    private let eye = SIMD3<Float>(0, 24, -46)
    private let motion: MotionTokens
    private let stats = FrameStats()
    private var press: Spring
    private var fade = 0.0
    private var time = 0.0
    private var goals = 0
    private var clockText = ""

    private var title = Entity()
    private var probe = Entity()
    private var button = ModelEntity()
    private var buttonMaterial = PhysicallyBasedMaterial()
    private var labelMaterial = PhysicallyBasedMaterial()
    private var buttonLabel = Entity()
    private var score: Entity?
    private var clock: Entity?

    init() throws {
        motion = try MotionTokens.load()
        press = Spring(motion.bouncy)
        try TextMesh.registerFont()
    }

    // MARK: building

    func build() async throws {
        // The world and its light as the match draws them (WorldStage): the spike's Oasis.
        let stage = try await WorldStage.load(.oasis, materials: try await Materials.load())
        root.addChild(stage.root)

        camera.camera.fieldOfViewInDegrees = fovDegrees
        camera.camera.fieldOfViewOrientation = .vertical
        camera.camera.near = 0.1
        camera.camera.far = 500
        camera.look(at: SIMD3(0, 0, 2), from: eye, relativeTo: nil)
        root.addChild(camera)

        buildTitle()
        score = replaceText(score, "0 : 0", height: 0.2, depth: 0.05, rgb: 0xFFFFFF, fromTop: 0.82)
        buildButton()
    }

    private func buildTitle() {
        title = TextMesh.entity(String(localized: "app.name"), height: 2.0, depth: 0.55, material: lit(0xFFF4D6))
        title.position = SIMD3(0, 4.5, 18)
        root.addChild(title)
        probe = TextMesh.entity(String(localized: "umlaut.probe"), height: 1.1, depth: 0.35, material: lit(0xFF6B6B))
        probe.position = SIMD3(0, 1.6, 17)
        probe.orientation = simd_quatf(angle: .pi, axis: [0, 1, 0])
        root.addChild(probe)
    }

    private func buildButton() {
        buttonMaterial = lit(0xFFC83D)
        labelMaterial = lit(0x1E1B4B)
        button = ModelEntity(mesh: .generateBox(width: 0.95, height: 0.34, depth: 0.16), materials: [buttonMaterial])
        button.position = SIMD3(0, -hudHalfHeight * 0.62, -hudDepth)
        button.components.set(DynamicLightShadowComponent(castsShadow: false))
        camera.addChild(button)
        buttonLabel = TextMesh.entity(buttonLabelText, height: 0.15, depth: 0.04, material: labelMaterial)
        buttonLabel.position = SIMD3(0, 0, 0.09)
        button.addChild(buttonLabel)
    }

    private var hudHalfHeight: Float { hudDepth * tan(fovDegrees / 2 * .pi / 180) }

    private func replaceText(_ old: Entity?, _ text: String, height: Float, depth: Float, rgb: Int, fromTop: Float) -> Entity {
        old?.removeFromParent()
        let node = TextMesh.entity(text, height: height, depth: depth, material: lit(rgb))
        node.position = SIMD3(0, hudHalfHeight * fromTop, -hudDepth)
        camera.addChild(node)
        return node
    }

    private func lit(_ rgb: Int) -> PhysicallyBasedMaterial {
        var m = PhysicallyBasedMaterial()
        m.baseColor = .init(tint: UIColor(red: CGFloat((rgb >> 16) & 0xFF) / 255, green: CGFloat((rgb >> 8) & 0xFF) / 255,
                                          blue: CGFloat(rgb & 0xFF) / 255, alpha: 1))
        m.roughness = 0.6
        m.metallic = 0.0
        return m
    }

    // MARK: frame

    /// The one clock (ADR 0005): called from the scene's update event, once per rendered frame.
    func update(_ dt: Double) {
        stats.record(dt)
        let step = min(dt, 0.1)
        time += step
        press.advance(step)
        if fade > 0 { fade = max(0, fade - step / motion.fadeSeconds) }

        title.position.y = 4.5 + 0.3 * Float(sin(time * 1.4))
        title.orientation = simd_quatf(angle: .pi + 0.12 * Float(sin(time * 0.9)), axis: [0, 1, 0])

        let s = Float(press.value)
        let bulge = 1 + Float(motion.pressBulge) * s
        button.scale = SIMD3(bulge, 1 - Float(motion.pressSquash) * s, bulge)
        let alpha = Float(1 - 0.6 * fade)
        setAlpha(alpha)

        let remaining = max(0, 120 - Int(time) % 121)
        let text = String(format: "%d:%02d", remaining / 60, remaining % 60)
        if text != clockText {
            clockText = text
            clock = replaceText(clock, text, height: 0.11, depth: 0.03, rgb: 0xFFE8A3, fromTop: 0.69)
        }
        project()
    }

    private var lastAlpha: Float = 1
    private func setAlpha(_ alpha: Float) {
        guard alpha != lastAlpha else { return }
        lastAlpha = alpha
        buttonMaterial.blending = alpha < 1 ? .transparent(opacity: .init(floatLiteral: alpha)) : .opaque
        labelMaterial.blending = buttonMaterial.blending
        button.model?.materials = [buttonMaterial]
        for case let model as ModelEntity in buttonLabel.children { model.model?.materials = [labelMaterial] }
    }

    // MARK: input and projection

    /// A touch on the view: our own hit-test against the 3D UI. True when the UI took it.
    @discardableResult
    func tap(at point: CGPoint) -> Bool {
        let (origin, dir) = ray(point)
        guard hit(button, origin: origin, dir: dir) else { return false }
        pressButton()
        return true
    }

    /// The button's action — reached by a tap on the 3D button or through the accessibility overlay.
    func pressButton() {
        if reduceMotion { fade = 1 } else { press.kick(motion.pressKick) }
        goals = (goals + 1) % 10
        score = replaceText(score, "\(goals) : 0", height: 0.2, depth: 0.05, rgb: 0xFFFFFF, fromTop: 0.82)
    }

    private var aspect: Float { Float(viewSize.width / max(viewSize.height, 1)) }

    private func ray(_ p: CGPoint) -> (SIMD3<Float>, SIMD3<Float>) {
        let ndcX = Float(2 * p.x / viewSize.width - 1)
        let ndcY = Float(1 - 2 * p.y / viewSize.height)
        let t = tan(fovDegrees / 2 * .pi / 180)
        let dirView = SIMD3<Float>(ndcX * t * aspect, ndcY * t, -1)
        let world = camera.transformMatrix(relativeTo: nil)
        let dir = simd_make_float3(world * SIMD4(dirView, 0))
        let origin = simd_make_float3(world.columns.3)
        return (origin, dir)
    }

    /// Ray against the entity's own-space bounds (the same slab test as Android's Node.hit).
    private func hit(_ e: Entity, origin: SIMD3<Float>, dir: SIMD3<Float>) -> Bool {
        let bounds = e.visualBounds(relativeTo: e)
        let o = e.convert(position: origin, from: nil)
        let d = e.convert(direction: dir, from: nil)
        var tMin: Float = 0, tMax = Float.greatestFiniteMagnitude
        for a in 0..<3 {
            if abs(d[a]) < 1e-8 {
                if o[a] < bounds.min[a] || o[a] > bounds.max[a] { return false }
            } else {
                let t1 = (bounds.min[a] - o[a]) / d[a], t2 = (bounds.max[a] - o[a]) / d[a]
                tMin = max(tMin, min(t1, t2)); tMax = min(tMax, max(t1, t2))
                if tMin > tMax { return false }
            }
        }
        return true
    }

    /// Projects the button's bounds to the screen for the accessibility overlay.
    private func project() {
        let b = button.visualBounds(relativeTo: button)
        let t = tan(fovDegrees / 2 * .pi / 180)
        var minX = CGFloat.greatestFiniteMagnitude, minY = CGFloat.greatestFiniteMagnitude
        var maxX = -CGFloat.greatestFiniteMagnitude, maxY = -CGFloat.greatestFiniteMagnitude
        for x in [b.min.x, b.max.x] { for y in [b.min.y, b.max.y] { for z in [b.min.z, b.max.z] {
            let world = button.convert(position: SIMD3(x, y, z), to: nil)
            let v = camera.convert(position: world, from: nil)
            guard v.z < 0 else { continue }
            let nx = (v.x / -v.z) / (t * aspect), ny = (v.y / -v.z) / t
            let sx = CGFloat((nx + 1) / 2) * viewSize.width, sy = CGFloat((1 - ny) / 2) * viewSize.height
            minX = min(minX, sx); maxX = max(maxX, sx); minY = min(minY, sy); maxY = max(maxY, sy)
        } } }
        let rect = minX < maxX ? CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY).integral : nil
        if rect != overlay.buttonRect { overlay.buttonRect = rect }
    }
}

File: android/app/src/main/kotlin/in/nann/smashhockey/spike/SpikeScene.kt
package `in`.nann.smashhockey.spike

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Typeface
import android.view.SurfaceView
import com.google.android.filament.Camera
import com.google.android.filament.Colors
import com.google.android.filament.ColorGrading
import com.google.android.filament.MaterialInstance
import com.google.android.filament.ToneMapper
import com.google.android.filament.View
import `in`.nann.smashhockey.R
import `in`.nann.smashhockey.engine.Assets
import `in`.nann.smashhockey.engine.FilamentHost
import `in`.nann.smashhockey.engine.FrameStats
import `in`.nann.smashhockey.engine.GpuMesh
import `in`.nann.smashhockey.engine.MeshBuilder
import `in`.nann.smashhockey.engine.MotionTokens
import `in`.nann.smashhockey.engine.Node
import `in`.nann.smashhockey.engine.Spring
import `in`.nann.smashhockey.engine.TextMesh
import `in`.nann.smashhockey.engine.World
import `in`.nann.smashhockey.core.generated.World as WorldId
import kotlin.math.PI
import kotlin.math.sin
import kotlin.math.tan

/** A screen rectangle in pixels, for the accessibility overlay. */
data class ScreenRect(val left: Int, val top: Int, val right: Int, val bottom: Int)

/**
 * The renderer spike's vertical slice (SMASH-2): the Oasis world, a floating title, a
 * camera-parented score HUD and one 3D button that squashes on a spring. Throwaway content on
 * the foundation it exists to prove — the engine package is what stays.
 */
class SpikeScene(
    private val context: Context,
    surfaceView: SurfaceView,
    private val onButtonRect: (ScreenRect?) -> Unit,
) {
    private val host = FilamentHost(surfaceView, ::frame)
    private val engine = host.engine
    private val assets = Assets(context.assets)
    private val motion = MotionTokens.load(assets)
    private val stats = FrameStats(context)
    private val typeface = Typeface.createFromAsset(context.assets, "fonts/LilitaOne-Regular.ttf")
    private val reduceMotion = !ValueAnimator.areAnimatorsEnabled()
    private val uiMaterial = assets.material(engine, "ui_lit")
    // The world and its light and fog as the match draws them (World): the spike's Oasis.
    private val world = World(engine, host.scene, host.view, assets, WorldId.OASIS)
    private val instances = mutableListOf<MaterialInstance>()
    private val meshes = mutableListOf<GpuMesh>()
    private val palette = mutableMapOf<Int, MaterialInstance>()
    private lateinit var buttonMaterial: MaterialInstance
    private lateinit var labelMaterial: MaterialInstance

    private val fovDegrees = 50.0
    private var aspect = 0.5
    private val hudDepth = 4f

    private lateinit var title: Node
    private lateinit var probe: Node
    private lateinit var button: Node
    private lateinit var buttonLabel: Node
    private var score: Node? = null
    private var clock: Node? = null
    private var clockText = ""
    private var goals = 0

    private val press = Spring(motion.bouncy)
    private var fade = 0.0
    private var time = 0.0
    private var lastRect: ScreenRect? = null

    val buttonLabelText: String = context.getString(R.string.play_button)

    init {
        host.view.colorGrading = ColorGrading.Builder().toneMapper(ToneMapper.Linear()).build(engine)
        host.view.dynamicResolutionOptions = View.DynamicResolutionOptions().apply { enabled = true; minScale = 0.6f }
        host.camera.setExposure(16f, 1f / 125f, 100f)
        host.camera.lookAt(0.0, 24.0, -46.0, 0.0, 0.0, 2.0, 0.0, 1.0, 0.0)
        host.onResize = { w, h ->
            aspect = w.toDouble() / h
            host.camera.setProjection(fovDegrees, aspect, 0.1, 500.0, Camera.Fov.VERTICAL)
            project()
        }

        buildTitle()
        buildHud()
        buildButton()
    }

    fun start() = host.start()
    fun stop() = host.stop()

    /** A tap on the surface: our own hit-test against the 3D UI. True when the UI took it. */
    fun tap(px: Float, py: Float): Boolean {
        val (origin, dir) = ray(px, py)
        return if (button.hit(origin, dir) != null || buttonLabel.hit(origin, dir) != null) { press(); true } else false
    }

    /** The button's action — reached by a tap on the 3D button or by the accessibility overlay. */
    fun press() {
        if (reduceMotion) fade = 1.0 else press.kick(motion.pressKick)
        goals = (goals + 1) % 10
        score = replaceText(score, "$goals : 0", 0.2f, 0.05f, 0xFFFFFF) { n -> hudPlace(n, 0.82f) }
    }

    // ---------------------------------------------------------------- frame

    private fun frame(dt: Double, @Suppress("UNUSED_PARAMETER") frameTimeNanos: Long) {
        stats.record(dt)
        val step = dt.coerceAtMost(0.1)
        time += step
        press.advance(step)
        if (fade > 0) fade = (fade - step / motion.fadeSeconds).coerceAtLeast(0.0)

        title.y = 4.5f + 0.3f * sin(time * 1.4).toFloat()
        title.yaw = PI.toFloat() + 0.12f * sin(time * 0.9).toFloat()
        title.apply()

        val s = press.value.toFloat()
        button.sy = 1f - motion.pressSquash.toFloat() * s
        button.sx = 1f + motion.pressBulge.toFloat() * s
        button.sz = button.sx
        button.apply()
        val alpha = (1.0 - 0.6 * fade).toFloat()
        buttonMaterial.setParameter("alpha", alpha)
        labelMaterial.setParameter("alpha", alpha)

        val remaining = (120 - time.toInt() % 121).coerceAtLeast(0)
        val text = "%d:%02d".format(remaining / 60, remaining % 60)
        if (text != clockText) {
            clockText = text
            clock = replaceText(clock, text, 0.11f, 0.03f, 0xFFE8A3) { n -> hudPlace(n, 0.69f) }
        }
        project()
    }

    // ---------------------------------------------------------------- building

    private fun buildTitle() {
        title = textNode(null, context.getString(R.string.app_name), 2.0f, 0.55f, 0xFFF4D6)
        title.z = 18f
        probe = textNode(null, context.getString(R.string.umlaut_probe), 1.1f, 0.35f, 0xFF6B6B)
        probe.x = 0f; probe.y = 1.6f; probe.z = 17f; probe.yaw = PI.toFloat(); probe.apply()
    }

    private fun buildHud() {
        score = replaceText(null, "0 : 0", 0.2f, 0.05f, 0xFFFFFF) { n -> hudPlace(n, 0.82f) }
    }

    private fun buildButton() {
        val block = MeshBuilder().box(0.95f, 0.34f, 0.16f).build()
        buttonMaterial = colour(0xFFC83D)
        labelMaterial = colour(0x1E1B4B)
        button = Node(engine, host.camera.entity, gpu(block, buttonMaterial))
        button.y = -hudHalfHeight() * 0.62f; button.z = -hudDepth
        button.apply()
        buttonLabel = Node(engine, button.entity, gpu(TextMesh.build(buttonLabelText, typeface, 0.15f, 0.04f), labelMaterial))
        buttonLabel.z = 0.09f
        buttonLabel.apply()
    }

    private fun hudHalfHeight() = (hudDepth * tan(Math.toRadians(fovDegrees / 2))).toFloat()

    private fun hudPlace(n: Node, fromTop: Float) {
        n.y = hudHalfHeight() * fromTop
        n.z = -hudDepth
        n.apply()
    }

    private fun textNode(parent: Int?, text: String, height: Float, depth: Float, rgb: Int): Node {
        val node = Node(engine, parent, gpu(TextMesh.build(text, typeface, height, depth), colour(rgb)))
        node.apply()
        return node
    }

    private fun replaceText(old: Node?, text: String, height: Float, depth: Float, rgb: Int, place: (Node) -> Unit): Node {
        old?.mesh?.let { host.scene.removeEntity(it.entity); meshes.remove(it); it.destroy() }
        val node = Node(engine, host.camera.entity, gpu(TextMesh.build(text, typeface, height, depth), colour(rgb)))
        place(node)
        return node
    }

    /** One material instance per colour, reused — text re-meshing must not leak instances. */
    private fun colour(rgb: Int): MaterialInstance = palette.getOrPut(rgb) {
        uiMaterial.createInstance().also { mi ->
            val c = Colors.toLinear(Colors.RgbType.SRGB, ((rgb shr 16) and 0xFF) / 255f, ((rgb shr 8) and 0xFF) / 255f, (rgb and 0xFF) / 255f)
            mi.setParameter("baseColor", Colors.RgbType.LINEAR, c[0], c[1], c[2])
            mi.setParameter("alpha", 1f)
            instances += mi
        }
    }

    private fun gpu(data: `in`.nann.smashhockey.engine.MeshData, mi: MaterialInstance): GpuMesh {
        val mesh = GpuMesh(engine, data, mi)
        meshes += mesh
        host.scene.addEntity(mesh.entity)
        return mesh
    }

    // ---------------------------------------------------------------- camera maths

    private fun ray(px: Float, py: Float): Pair<FloatArray, FloatArray> {
        val ndcX = 2.0 * px / host.width - 1.0
        val ndcY = 1.0 - 2.0 * py / host.height
        val t = tan(Math.toRadians(fovDegrees / 2))
        val dv = doubleArrayOf(ndcX * t * aspect, ndcY * t, -1.0)
        val m = DoubleArray(16).also { host.camera.getModelMatrix(it) }
        val dir = FloatArray(3) { i -> (m[i] * dv[0] + m[4 + i] * dv[1] + m[8 + i] * dv[2]).toFloat() }
        val origin = floatArrayOf(m[12].toFloat(), m[13].toFloat(), m[14].toFloat())
        return origin to dir
    }

    /** Projects the button's bounds to the screen for the accessibility overlay. */
    private fun project() {
        if (!::button.isInitialized) return
        val view = DoubleArray(16).also { host.camera.getViewMatrix(it) }
        val proj = DoubleArray(16).also { host.camera.getProjectionMatrix(it) }
        var l = Float.MAX_VALUE; var t = Float.MAX_VALUE; var r = -Float.MAX_VALUE; var b = -Float.MAX_VALUE
        for (p in button.worldCorners()) {
            val v = mul(view, p); val c = mul(proj, v)
            if (c[3] <= 0) continue
            val sx = ((c[0] / c[3] + 1) / 2 * host.width).toFloat()
            val sy = ((1 - c[1] / c[3]) / 2 * host.height).toFloat()
            l = minOf(l, sx); r = maxOf(r, sx); t = minOf(t, sy); b = maxOf(b, sy)
        }
        val rect = if (l < r) ScreenRect(l.toInt(), t.toInt(), r.toInt(), b.toInt()) else null
        if (rect != lastRect) { lastRect = rect; onButtonRect(rect) }
    }

    private fun mul(m: DoubleArray, p: FloatArray): DoubleArray = DoubleArray(4) { i ->
        m[i] * p[0] + m[4 + i] * p[1] + m[8 + i] * p[2] + m[12 + i] * (if (p.size > 3) p[3].toDouble() else 1.0)
    }

    private fun mul(m: DoubleArray, p: DoubleArray): DoubleArray = DoubleArray(4) { i ->
        m[i] * p[0] + m[4 + i] * p[1] + m[8 + i] * p[2] + m[12 + i] * p[3]
    }

    fun destroy() {
        host.stop()
        meshes.forEach { it.destroy() }
        instances.forEach { engine.destroyMaterialInstance(it) }
        engine.destroyMaterial(uiMaterial)
        world.destroy()
        host.destroy()
    }
}
